proposals/utils/stepper.py

```diff
@@ -0,0 +1,157 @@
+from django.utils.translation import gettext as _
+
+from main.utils import renderable
+
+from .stepper_helpers import (
+    PlaceholderItem,
+    StepperItem,
+)
+
+from .checkers import ProposalTypeChecker
+
+
+class Stepper(renderable):
+
+    template_name = "base/stepper.html"
+
+    def __init__(
+        self,
+        proposal,
+        proposal_type_hint=None,
+        request=None,
+    ):
+        self.proposal = proposal
+        self.starting_checkers = [
+            ProposalTypeChecker,
+        ]
+        # The stepper keeps track of the request to determine
+        # which item is current
+        self.request = request
+        # The type can be provided by the view for the case in which
+        # the proposal has not yet been created but we need to determine
+        # its type, i.e. the ProposalCreateViews
+        self.proposal_type_hint = proposal_type_hint
+        self.items = []
+        self.check_all(self.starting_checkers)
+
+    def get_context_data(self):
+        context = super().get_context_data()
+        # Provide the stepper bubble classes in order
+        # of descending size
+        bubble_list = [
+            "stepper-bubble-largest",
+            "stepper-bubble-large",
+            "stepper-bubble-medium",
+            "stepper-bubble-small",
+            "stepper-bubble-smallest",
+        ]
+        context.update(
+            {
+                "stepper": self,
+                "bubble_size": bubble_list,
+            }
+        )
+        return context
+
+    def get_resume_url(self):
+        """
+        Returns the url of the first page that requires attention,
+        that being either a page with an error or an incomplete page.
+        """
+        for item in self.items:
+            if not item.is_complete:
+                return item.get_url()
+
+    def build_stepper(
+        self,
+    ):
+        """
+        The meat and potatoes of the stepper. Returns a list of top-level
+        StepperItems to be rendered in the template.
+        """
+        layout = getattr(self, "layout", False)
```

Comment:
I don't think so, because in almost all cases the Stepper can determine its type (and thereby layout) all by itself. The createviews are the only exception I can imagine. So if the layout was an argument it definitely should not be required.

Second, the concept of a "layout" is completely internal to the stepper implementation. I don't think every View that instantiates a stepper should need to know what a layout is and when to choose which one. The createviews know what *type of proposal* they're creating, because they already have to, and that should be good enough.

I think this is important because there may not exist a 1:1 mapping between proposal types and layouts. A practice proposal might use the same layout as a regular proposal, but require a different checker flow because it asks for a practice reason. Because a checker can do both of those things with just the proposal type, it makes sense to me to put that in a `ProposalTypeChecker` which then also determines the layout.

Adding to the above, although it isn't necessarily relevant to your comment, is that I think that the state of the stepper object should only be modified by Checkers. Once an object is complicated enough to warrant this kind of crazy depth-first search operation with Checkers, I want to respect that and not also change object state in various object methods.

Otherwise, I can imagine myself stepping through the checkers for ages trying to debug something, only to find that it was in fact that one thing I decided to put in a Stepper method instead.